Route resolvers

Allcharles opened the pull request:
# Route Resolvers 2.0

Re-implemented resolvers to attach them to a page instead of the category. This removed inheritance of resolvers from parent pages as well.

## Changes

- Fixed child categories redefining parent routes
- Flattened routes instead of previous nested route solution
- Removed category resolvers
- Updated app dependencies

## Closes

Closes #159 
Closes #158

Allcharles committed:
Updated dependencies

Allcharles committed:
Updating route resolvers.
List of changes:
- Updated dependencies
- Handle child category bug
- Flattened router paths so that there are no more nested children
- Extracted resolvers to pages

Allcharles committed:
Handle child category bug

Initial solution to issue #159. Currently breaks the route path for child categories however

Allcharles committed:
Change route paths

- Flattened router paths so that there are no more nested children
- Extracted resolvers to pages

Allcharles committed:
Fixed home component route

Allcharles committed:
Fixed unit tests

Allcharles committed:
Added route sorting

Fixed bug where attribute routes gain priority over non-attribute routes.

Allcharles committed:
Merge branch 'route-resolvers' of https://github.com/Allcharles/AngularTestbench into route-resolvers

atruskie commented on src/app/interfaces/strongRoute.ts:
```diff
@@ -66,9 +76,10 @@ export class StrongRoute {
    */
-  add(name: string, config: Partial<Route> = {}) {
-    return new StrongRoute(this, name, config);
+  add(name: string, config: Partial<Route> = {}, isRoot?: boolean) {
```
why are we exposing `isRoot` publically?

Allcharles replied:
Maybe the variable name is a little misleading, it is designed for child categories such as the Sites Category which allows it to set itself as the root, even though the Projects Category is the actual root. Basically it lets me stop the child category redefining the parent category routes and vice versa. This is part of the solution for #159

atruskie replied:
Responded to #159

atruskie reviewed (approved):
It's a little worrying that all the resolvers are referred to by strings.

Allcharles committed:
Address review comments

Allcharles committed:
Rename addNested to addFeatureModule

Allcharles committed:
Extract resolver keys to variable

atruskie commented on src/app/interfaces/strongRoute.ts:
```diff
@@ -76,10 +76,18 @@ export class StrongRoute {
+   * @param name Route name
+   * @param config Additional router configurations
+   */
+  addFeatureModule(name: string, config: Partial<Route> = {}) {
```
As per https://github.com/QutEcoacoustics/workbench-client/issues/159#issuecomment-603528912 I'd prefer if mature models used a different flag and did not reuse the `isRoot` property.

This can be fixed in a later commit, however.

Allcharles committed:
Resolve review comment

Addressing https://github.com/QutEcoacoustics/workbench-client/issues/159#issuecomment-603528912
